### PR TITLE
Drop leap days based on CF calendar type to calculate daily climatologies and departures 

### LINKED ISSUE #303: [Doc]: Calculating Daily Climatology and Departures from Monthly Time Series needs to use daily input
### Describe your documentation update

https://xcdat.readthedocs.io/en/v0.3.0/examples/climatology-and-departures.html#2.4-Daily-Cycle

I think the example calculation of daily climatology did not work properly because the input data was monthly data. It is expected that the output time steps are 365 or 366, but the example output time steps are 12.

### LINKED ISSUE #337: [Bug]: Drop leap years for high frequency input data when calculating climatology/departures
### What happened?
Related to https://github.com/xCDAT/xcdat/issues/303

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [27], in <cell line: 1>()
----> 1 cftime.DatetimeProlepticGregorian(1, 2, 29, 0, 0, 0, 0, has_year_zero=True)

File src/cftime/_cftime.pyx:2016, in cftime._cftime.DatetimeProlepticGregorian.__init__()

File src/cftime/_cftime.pyx:1145, in cftime._cftime.datetime.__init__()

File src/cftime/_cftime.pyx:1647, in cftime._cftime.assert_valid_date()

ValueError: invalid day number provided in cftime.DatetimeProlepticGregorian(1, 2, 29, 0, 0, 0, 0, has_year_zero=True)

```

### What did you expect to happen?

xCDAT is able to handle leap years when calculating climatology/departures for high frequency data.

### Proposed Solution

Discussed in https://github.com/xCDAT/xcdat/discussions/332

<div type='discussions-op-text'>

<sup>Originally posted by **lee1043** August 31, 2022</sup>
Following https://github.com/xCDAT/xcdat/issues/303#issuecomment-1218705475, calculation of climatology for daily or higher temporal frequency time series will need to remove data on dates of 29th February in leap years from the input time series. 

Calendars having normal leap year in CF-convention are "gregorian", "standard", and "proleptic_gregorian". (Calendars in CF-convention: https://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch04s04.html)

We may want to implement a kind of pre-processing function for input data, which could be something like below. I am yet to have a good idea where this function should be placed. @tomvothecoder would you share your insight?

```python
# Get daily climatology (ds: dataset with daily time series)
ds_daily_climo = ds.temporal.climatology("tas", freq="day", weighted=True)

# Below is somewhere in the temporal.climatology, in advance to calculate climatology, to pre-process input
def function(ds, freq):
    if freq in ["day", "hour"]:
        if calendar in ["gregorian", "standard", "proleptic_gregorian"]:
            # Remove leap year prior to daily climatology
            ds2 = ds.sel(time=~((ds.time.dt.month==2)&(ds.time.dt.day==29)))
            ds = copy.deepcopy(ds2)
    return ds
```
</div>

### Minimal Complete Verifiable Example
```python
 import cftime
cftime.DatetimeProlepticGregorian(0, 2, 29, 0, 0, 0, 0, has_year_zero=True)  # okay (leap year every 4th year starting 0th)
cftime.DatetimeProlepticGregorian(1, 2, 29, 0, 0, 0, 0, has_year_zero=True)  # Error (because no leap year in 1th year)
```

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by tomvothecoder**
## Description

Summary of Changes
- Closes #337 
- Closes #332 
- Closes #303 
- Includes additional refactoring to keep time bounds variable deterministic in the temporal averaging accessor class.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass with my changes (locally and CI/CD build)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: xcdat/temporal.py [hunk 1]
```diff
@@ -816,75 +904,52 @@ def _drop_incomplete_djf(self, dataset: xr.Dataset) -> xr.Dataset:
...
-                    f"Supported months include: {predefined_months}."
-                )
-
-        c_seasons = {"".join(months): months for months in custom_seasons}
-
-        return c_seasons
+        ds = ds.sel(  # type: ignore
+            **{self._dim: ~((ds.time.dt.month == 2) & (ds.time.dt.day == 29))}
+        )
+        return ds
```
Comment (tomvothecoder) [inline xcdat/temporal.py:940]: This is the new method to handle dropping leap years when calculating daily climatology/departures.

File: xcdat/temporal.py [hunk 2]
```diff
@@ -784,6 +790,88 @@ def _set_obj_attrs(
...
+            and self._season_config.get("drop_incomplete_djf") is True
+        ):
+            ds = self._drop_incomplete_djf(ds)
+
+        if (
+            self._freq == "day"
+            and self._mode in ["climatology", "departures"]
+            and self.calendar in ["gregorian", "proleptic_gregorian", "standard"]
+        ):
+            ds = self._drop_leap_days(ds)
```
Comment (tomvothecoder) [inline xcdat/temporal.py:871]: The logic to check whether or not to drop leap years.

File: xcdat/temporal.py [hunk 3]
```diff
@@ -516,7 +518,10 @@ def departures(
 
             * "season": groups by season for the seasonal cycle departures.
             * "month": groups by month for the annual cycle departures.
-            * "day": groups by (month, day) for the daily cycle departures.
+            * "day": groups by (month, day) for the daily cycle departures. Leap
+              days (if present) are dropped if the CF calendar type is
+              ``"gregorian"``, ``"proleptic_gregorian"``, or ``"standard"``
```
Comment (tomvothecoder) [inline xcdat/temporal.py:523]: Docstring update to inform users when leap days are dropped.

File: xcdat/temporal.py [hunk 4]
```diff
@@ -784,6 +790,88 @@ def _set_obj_attrs(
         else:
             self._season_config["custom_seasons"] = self._form_seasons(custom_seasons)
 
+    def _form_seasons(self, custom_seasons: List[List[str]]) -> Dict[str, List[str]]:
```
Comment (tomvothecoder) [inline xcdat/temporal.py:793]: `_form_seasons` was simply reordered, no changes were made.

File: tests/test_temporal.py [hunk 5]
```diff
@@ -1382,6 +1383,97 @@ def test_weighted_daily_climatology(self):
 
         assert result.identical(expected)
 
+    def test_weighted_daily_climatology_drops_leap_days_with_matching_calendar(self):
```
Comment (tomvothecoder) [inline tests/test_temporal.py:1386]: New test for daily climatology.

File: tests/test_temporal.py [hunk 6]
```diff
@@ -1606,6 +1698,126 @@ def test_unweighted_seasonal_departures_with_JFD(self):
 
         assert result.identical(expected)
 
+    def test_weighted_daily_departures_drops_leap_days_with_matching_calendar(self):
```
Comment (tomvothecoder) [inline tests/test_temporal.py:1701]: New test for daily departures.

File: tests/test_temporal.py [hunk 7]
```diff
@@ -1654,7 +1866,7 @@ def test_weights_for_yearly_averages(self):
             )
 
             # Compare result of the method against the expected.
-            result = ds.temporal._get_weights()
+            result = ds.temporal._get_weights(ds.time_bnds)
```
Comment (tomvothecoder) [inline tests/test_temporal.py:1869]: Ignore these changes to `_get_weights()` (result of refactoring).

**Review by tomvothecoder — COMMENTED** [review]
Hi @lee1043 and @msahn this PR is ready for review. I commented areas of interest, and you can ignore other changes because they are mostly refactoring.

The climatology and departures notebook has been fixed to use hourly data for hourly calculations. The changes can be previewed here:
https://xcdat.readthedocs.io/en/feature-337-climo-highfreq/examples/climatology-and-departures.html#

Thanks!

**Review by lee1043 — APPROVED** [review]
I think this PR can be merged. @tomvothecoder thank you for the update!

**Pull request merged**